learning-another-language/js-loops.md

```diff
@@ -0,0 +1,417 @@
+
+While you can use the `for … in` to iterate over an array, it is recommended to use one of the aforementioned for-loops instead. 
+
+The reason for this is because the `for … in` loop iterates over **all** of the enumerable properties in an object, including any that may be inherited.
+
+If, for instance, a JS library modifies the `Array` prototype, the loop will iterate over any additional properties as well. Check out the example below for a potentially non-intuitive demonstration.
+
+```js
+const arr = [1, 2, 3];
+
+Array.prototype.newMethod = true;
+
+for (const i in arr) {
+	console.log(i);
+}
+
+// Output:
+// 1
+// 2
+// 3
+// newMethod
+```
+
+It is also worth mentioning that the `for … in` loop is meant for objects, and thus will be slower than other loops which are better suited for arrays.
+
+### Practical Usage
+
+So, wait, why use the `for … in` loop at all?
+
+The `for … in` loop may be most practically used as a tool to debug. It is an easy way to check the properties of an object. 
+
+## 4. while loop
+```js
+while (condition) {
+	// do some repetitive task
+}
+```
+
+The `while` loop will evaluate the `condition` before the loop is run each time. If the `condition` evaluates to `true`, the loop executes the statement(s) in the block. Otherwise, the loop exits and stops executing the statement(s) in the block.
+
+The `while` loop is known as a pre-test loop because it evaluates the `condition` *before* each iteration. The loop will never execute if the condition evaluates to `false` before entering the loop.
+
+### Example
+
+```js
+let count = 5;
+while (count > 0) {
+	console.log(count);
+	count -= 1;
+}
+
+// Output:
+// 5
+// 4
+// 3
+// 2
+// 1
+```
+
+Try converting the following `for-loop` into a `while` loop:
+```js
+for (int i = 0; i <= 20; i += 5) {
+	console.log(i);
+}
+```
+
+<details style="max-width: 700px; margin: auto;">
+  <summary>
+  Click here to reveal a possible solution
+  </summary>
+
+  ```javascript
+    let i = 0;
+    while (i <= 20) {
+        console.log(i);
+        i += 5;
+    }
+
+    // Output:
+    // 0
+    // 5
+    // 10
+    // 15
+    // 20
+  ```
+</details>
+
+## 5. do…while loop
+```js
+do {
+	// perform repetitive task at least once
+} while(condition)
+```
+
+The `do...while` statement creates a loop that executes a block of code until a condition evaluates to `false`.
+
+The statements in the block will **always** be executed once before the condition is checked. This is what differentiates a `do...while` loop from a `while` loop. The `do...while` loop is sometimes called a post-test loop.
+
+### Example
+```js
+let i = 1
+do {
+	console.log(i);
+	i += 1;
+} while (i <= 5)
+
+// Output:
+// 1
+// 2
+// 3
+// 4
+// 5
+```
+
+### Common Pitfall: Indefinite Loop
+
+It’s important to ensure the variable(s) used in the statement are updated such that the `while(condition)` evaluates to false. Otherwise, you may crash your program due to an infinite loop.
```

Comment:
```suggestion
It’s important to ensure the variable(s) used in the condition are updated such that the `while(condition)` evaluates to false. Otherwise, our program may crash due to an infinite loop.
```